website/src/react-luma/components/luma-example.tsx

```diff
@@ -63,52 +68,78 @@ export const LumaExample: FC<LumaExampleProps> = (props: LumaExampleProps) => {
   let containerName = 'ssr';
 
   /** Each example maintains an animation loop */
-  const [animationLoop, setAnimationLoop] = useState<AnimationLoop | null>(null);
+  const animationLoopRef = useRef<AnimationLoop | null>(null);
+  const [canvas, setCanvas] = useState<HTMLCanvasElement | null>(null);
+  const usedCanvases = useRef(new WeakMap())
+  const currentTask = useRef<Promise<any> | null>(null);
 
   /** Type type of the device (WebGL, WebGPU, ...) */
   const deviceType = useStore(store => store.deviceType);
   containerName = props.container || `luma-example-container-${deviceType}`;
 
-  const callbackRef = useCallback((canvas: HTMLCanvasElement) => {
-    if (canvas) {
-      if (!animationLoop) {
-        log.info(0, `creating luma device ${canvas.id}`); // , ref.current);
-        canvas.style.width = '100%';
-        canvas.style.height = '50%';
-        const device = luma.createDevice({type: deviceType, canvas, container: containerName});
-
-        let animationLoop: AnimationLoop | null = null;
-        animationLoop = makeAnimationLoop(props.template as unknown as typeof AnimationLoopTemplate, {
-          device,
-          autoResizeViewport: true,
-          autoResizeDrawingBuffer: true
-        });
 
-        // Start the actual example
-        animationLoop?.start();
-        setAnimationLoop(animationLoop);
-
-        // Ensure the example can find its images
-        // TODO - this only works for examples/tutorials
-        const RAW_GITHUB = 'https://raw.githubusercontent.com/visgl/luma.gl/master';
-        if (props.directory) {
-          setPathPrefix(`${RAW_GITHUB}/examples/${props.directory}/${props.id}/`);
-        } else {
-          setPathPrefix(`${RAW_GITHUB}/website/static/images/`);
-        }
-      }
+  useEffect(() => {
+    if (!canvas || usedCanvases.current.get(canvas)) return;
+
+    usedCanvases.current.set(canvas, true)
+
+    let animationLoop: AnimationLoop | null = null;
+    let device: Device | null = null;
+    const asyncCreateLoop = async () => {
+      if (animationLoopRef.current) return;
+
+      console.log(`starting example ${deviceType}`);
+      canvas.style.width = '100%';
+      canvas.style.height = '100%';
+      device = await luma.createDevice({type: deviceType, canvas, container: containerName});
 
-    } else {
+      animationLoop = makeAnimationLoop(props.template as unknown as typeof AnimationLoopTemplate, {
```

Comment:
Nit: put such a nasty cast on its own line?

```suggestion
      const animationLoopTemplate = props.template as unknown as typeof AnimationLoopTemplate;
      animationLoop = makeAnimationLoop(animationLoopTemplate, {
```